### ISSUE TITLE
flatpak-cargo-builder fails because Cargo.toml in git submodule is not found

### ISSUE THREAD

**Issue by mfschumann** [flatpak/flatpak-builder-tools#322]
### flatpak-builder version

1.2.2

### Linux distribution and version

Fedora Silverblue 36

### Affected flatpak-builder tool

cargo/flatpak-cargo-generator.py

### flatpak-builder tool cli args

Cargo.lock

### Source repository URL

https://github.com/flathub/eu.betterbird.Betterbird/tree/58b020362586ad089249282a3e019705f310f8ff

### Description

I got the following error when trying to regenerate a sources json from an updated Cargo.lock for eu.betterbird.Betterbird:
```
Traceback (most recent call last):
  File "/var/home/martin/Projekte/eu.betterbird.Betterbird/flatpak-builder-tools/cargo/flatpak-cargo-generator.py", line 356, in <module>
    main()
  File "/var/home/martin/Projekte/eu.betterbird.Betterbird/flatpak-builder-tools/cargo/flatpak-cargo-generator.py", line 349, in main
    generated_sources = asyncio.run(generate_sources(load_toml(args.cargo_lock),
  File "/usr/lib64/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib64/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/var/home/martin/Projekte/eu.betterbird.Betterbird/flatpak-builder-tools/cargo/flatpak-cargo-generator.py", line 303, in generate_sources
    for pkg in await asyncio.gather(*pkg_coros):
  File "/var/home/martin/Projekte/eu.betterbird.Betterbird/flatpak-builder-tools/cargo/flatpak-cargo-generator.py", line 256, in get_package_sources
    return await get_git_package_sources(package, git_repos)
  File "/var/home/martin/Projekte/eu.betterbird.Betterbird/flatpak-builder-tools/cargo/flatpak-cargo-generator.py", line 203, in get_git_package_sources
    git_repo['commits'][commit] = await get_git_repo_packages(repo_url, commit)
  File "/var/home/martin/Projekte/eu.betterbird.Betterbird/flatpak-builder-tools/cargo/flatpak-cargo-generator.py", line 161, in get_git_repo_packages
    await get_dep_packages(pkg_toml, subpkg)
  File "/var/home/martin/Projekte/eu.betterbird.Betterbird/flatpak-builder-tools/cargo/flatpak-cargo-generator.py", line 143, in get_dep_packages
    dep_toml = load_toml(os.path.join(git_repo_dir, dep_dir, 'Cargo.toml'))
  File "/var/home/martin/Projekte/eu.betterbird.Betterbird/flatpak-builder-tools/cargo/flatpak-cargo-generator.py", line 99, in load_toml
    with open(tomlfile, 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/var/home/martin/.cache/flatpak-cargo/https_github.com_mozilla_application-services/components/external/glean/glean-core/ffi/Cargo.toml'
```

The folder `/var/home/martin/.cache/flatpak-cargo/https_github.com_mozilla_application-services/components/external` only contains `glean`, which is empty. 

The issue seems to be caused by a submodule in the https://github.com/mozilla/application-services repo not being checked out when cloning the repo without `--recursive`: https://github.com/flatpak/flatpak-builder-tools/blob/4cdd0308f50f5980fda7e86c402c7a998352f52e/cargo/flatpak-cargo-generator.py#L114

Also, the `exclude` list in [Cargo.toml at mozilla/application-services](https://github.com/mozilla/application-services/blob/953c414612f379c2b1a0b806cf6416d079a1c3a6/Cargo.toml#L44) seems to be ignored by flatpak-cargo-generator, causing the missing submodule issue to surface in the first place.

**Comment by gasinvein** [comment]
Thanks for detailed issue. You seem more proficient with Cargo then I am; what do you think would be the optimal solution here - clone git repos recursively, acknowledge excluded paths, or both?

**Comment by mfschumann** [comment]
Well, for my particular situation acknowledging the excluded paths should be enough. But I can imagine that cloning recursively could be needed in other situations.

**Comment by gasinvein** [comment]
Hmm, I'm not sure I understand how `exclude` paths are handled by Cargo. The docs say
>The `exclude` key can be used to prevent paths from being included in a workspace. This can be useful if some path dependencies aren't desired to be in the workspace at all, or using a glob pattern and you want to remove a directory.

So, is a path listed in `workspace.exclude` excluded only from workspace members, or from member package dependencies?
In this specific case, `application-services` has `megazords/ios/rust` member, which, in turn, has `components/external/glean/glean-core/ffi` dependency, which is listed in the workspace `exclude`. So, should we just ignore this dependency? It doesn't feel right.

**Comment by mfschumann** [comment]
From [The Cargo Book](https://doc.rust-lang.org/cargo/reference/workspaces.html):
> [[workspace]](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-workspace-section) — Defines a workspace.
[members](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields) — Packages to include in the workspace.
[exclude](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-members-and-exclude-fields) — Packages to exclude from the workspace.

To me it seems that `exclude` does the opposite of `members`, i.e. any packages listed in `exclude` will not be added to the workspace. The text you quoted indicates that this applies both to packages that got selected via a glob pattern in `members` and to dependencies. At least that is my understanding of the documentation; I am no Cargo expert either.

**Comment by gasinvein** [comment]
@mfschumann Can you please try `flatpak-cargo-generator.py` from #323 and tell if it works for you (including at least `cargo --offline fetch` passing at build time, or complete build)?

**Comment by mfschumann** [comment]
With this version the error does not occur anymore. Thanks for the quick fix!

**Comment by gasinvein** [comment]
Did you manage to build the app (or at least get Cargo past "fetch" stage and start compiling)?

**Comment by mfschumann** [comment]
Well, not really. The error in `flatpak-cargo-generator.py` was gone, but the generated `.json` produced an error during Cargo fetch. I then realized that I was using the wrong `Cargo.lock` for my project, so I am quite confident that this mistake is the actual cause for the Cargo error I was getting after your bugfix. 

The correct `Cargo.lock` for my project does not provoke the issue I reported here in the first place, because it does not use submodules or `exclude` definitions. Building my project works fine when using a `.json` generated using `flatpak-cargo-generator.sh` from this PR and the correct `Cargo.lock`.
